### PR TITLE
Refactoring interpreter and paranoid mode, introducing traits to allo…

### PR DESCRIPTION
**Pull request by ziaptos**
## Description

Refactoring interpreter and paranoid mode, introducing traits to allow generic interpreter as well as compile time switch for the runtime type checks (formerly called paranoid mode)

## How Has This Been Tested?
Existing tests

## Key Areas to Review
Should focus on making sure that the high level logic has *not* changed!

## Type of Change
- [ ] New feature
- [ ] Bug fix
- [ ] Breaking change
- [x] Performance improvement
- [x] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [x] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Move Compiler
- [ ] Other (specify)

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: third_party/move/move-vm/runtime/src/runtime_type_checks.rs [hunk 1]
```diff
@@ -0,0 +1,683 @@
...
+    file_format::{Ability, AbilitySet, Bytecode},
+};
+use move_core_types::vm_status::StatusCode;
+use move_vm_types::{loaded_data::runtime_types::Type, values::Locals};
+
+pub(crate) trait RuntimeTypeCheck {
+    fn pre_execution_type_stack_transition(
+        local_tys: &[Type],
+        locals: &Locals,
+        _ty_args: &[Type],
```
Comment (georgemitenkov) [inline third_party/move/move-vm/runtime/src/runtime_type_checks.rs:16]: `ty_args` and `resolver`

File: third_party/move/move-vm/runtime/src/runtime_type_checks.rs [hunk 2]
```diff
@@ -0,0 +1,683 @@
...
+        // For example, it is ok to have a struct that doesn't have a copy capability where its field is a struct that has copy capability but not vice versa.
+        ty.paranoid_check_abilities(field_expected_abilities)?;
+        ty.paranoid_check_eq(expected_ty)?;
+    }
+
+    operand_stack.push_ty(output_ty)
+}
+
+pub(crate) struct NullRuntimeTypeCheck;
+pub(crate) struct FullRuntimeTypeCheck;
```
Comment (georgemitenkov) [inline third_party/move/move-vm/runtime/src/runtime_type_checks.rs:66]: nit: move `FullRuntimeTypeCheck` to trait impl? Also add brief documentation? Space between struct definitions?

File: third_party/move/move-vm/runtime/src/runtime_type_checks.rs [hunk 3]
```diff
@@ -0,0 +1,683 @@
...
+        // Fields ability should be a subset of the struct ability because abilities can be weakened but not the other direction.
+        // For example, it is ok to have a struct that doesn't have a copy capability where its field is a struct that has copy capability but not vice versa.
+        ty.paranoid_check_abilities(field_expected_abilities)?;
+        ty.paranoid_check_eq(expected_ty)?;
+    }
+
+    operand_stack.push_ty(output_ty)
+}
+
+pub(crate) struct NullRuntimeTypeCheck;
```
Comment (georgemitenkov) [inline third_party/move/move-vm/runtime/src/runtime_type_checks.rs:65]: See some C++ here ahah, `NoRuntimeTypeCheck` maybe?

File: third_party/move/move-vm/runtime/src/runtime_type_checks.rs [hunk 4]
```diff
@@ -0,0 +1,683 @@
...
+            | Bytecode::MutBorrowVariantFieldGeneric(_)
+            | Bytecode::ImmBorrowVariantField(_)
+            | Bytecode::ImmBorrowVariantFieldGeneric(_) => (),
+        };
+        Ok(())
+    }
+
+    /// Paranoid type checks to perform after instruction execution.
+    ///
+    /// This function and `pre_execution_type_stack_transition` should constitute the full type stack transition for the paranoid mode.
```
Comment (georgemitenkov) [inline third_party/move/move-vm/runtime/src/runtime_type_checks.rs:217]: I think lines are > 100 chars again?

File: third_party/move/move-vm/runtime/src/runtime_type_checks.rs [hunk 5]
```diff
@@ -0,0 +1,683 @@
...
+        let ty_builder = resolver.loader().ty_builder();
+
+        match instruction {
+            Bytecode::BrTrue(_) | Bytecode::BrFalse(_) => (),
+            Bytecode::Branch(_)
+            | Bytecode::Ret
+            | Bytecode::Call(_)
+            | Bytecode::CallGeneric(_)
+            | Bytecode::Abort => {
+                // Invariants hold because all of the instructions above will force VM to break from the interpreter loop and thus not hit this code path.
```
Comment (georgemitenkov) [inline third_party/move/move-vm/runtime/src/runtime_type_checks.rs:235]: again, multiple lines

File: third_party/move/move-vm/runtime/src/runtime_type_checks.rs [hunk 6]
```diff
@@ -0,0 +1,683 @@
...
+        local_tys: &[Type],
+        ty_args: &[Type],
+        resolver: &Resolver,
+        operand_stack: &mut Stack,
+        ty_cache: &mut FrameTypeCache,
+        instruction: &Bytecode,
+    ) -> PartialVMResult<()>;
+}
+
+/// Paranoid type checks to perform before instruction execution.
```
Comment (georgemitenkov) [inline third_party/move/move-vm/runtime/src/runtime_type_checks.rs:32]: If this is meant for the module, you can do `//! Your description here` on top of the file? This also seems like a good documentation to add to trait definition?

File: third_party/move/move-vm/runtime/src/runtime_type_checks.rs [hunk 7]
```diff
@@ -0,0 +1,683 @@
...
+            .union(AbilitySet::singleton(Ability::Store))
+    } else {
+        ability
+    };
+    for (ty, expected_ty) in operand_stack
+        .popn_tys(field_count)?
+        .into_iter()
+        .zip(field_tys)
+    {
+        // Fields ability should be a subset of the struct ability because abilities can be weakened but not the other direction.
```
Comment (georgemitenkov) [inline third_party/move/move-vm/runtime/src/runtime_type_checks.rs:56]: Lines too long

File: third_party/move/move-vm/runtime/src/interpreter.rs [hunk 8]
```diff
@@ -98,33 +110,76 @@ impl Interpreter {
...
+    }
 
     /// Main loop for the execution of a function.
     ///
     /// This function sets up a `Frame` and calls `execute_code_unit` to execute code of the
     /// function represented by the frame. Control comes back to this function on return or
     /// on call. When that happens the frame is changes to a new one (call) or to the one
     /// at the top of the stack (return). If the call stack is empty execution is completed.
-    fn execute_main(
+    fn execute_main<RTTCheck: RuntimeTypeCheck>(
```
Comment (georgemitenkov) [inline third_party/move/move-vm/runtime/src/interpreter.rs:182]: Love the generic bound name :)

File: third_party/move/move-vm/runtime/src/interpreter.rs [hunk 9]
```diff
@@ -1147,7 +1204,7 @@ const CALL_STACK_SIZE_LIMIT: usize = 1024;
 pub(crate) const ACCESS_STACK_SIZE_LIMIT: usize = 256;
 
 /// The operand stack.
```
Comment (georgemitenkov) [inline third_party/move/move-vm/runtime/src/interpreter.rs:1206]: nit: "operand and type"?

File: third_party/move/move-vm/runtime/src/interpreter.rs [hunk 10]
```diff
@@ -1117,9 +1151,32 @@ impl Interpreter {
     fn get_internal_state(&self) -> ExecutionState {
         self.get_stack_frames(usize::MAX)
     }
+}
+
+impl InterpreterDebugInterface for InterpreterImpl {
+    #[allow(dead_code)]
```
Comment (georgemitenkov) [inline third_party/move/move-vm/runtime/src/interpreter.rs:1157]: Can we use some `cfg(any(...))` or so? Seems like this is only used if some cfg feature is on

File: third_party/move/move-vm/runtime/src/interpreter.rs [hunk 11]
```diff
@@ -98,33 +110,76 @@ impl Interpreter {
...
+    ) -> VMResult<Vec<Value>> {
+        let interpreter = InterpreterImpl {
+            operand_stack: Stack::new(),
+            call_stack: CallStack::new(),
+            paranoid_type_checks: loader.vm_config().paranoid_type_checks,
+            access_control: AccessControlState::default(),
+            active_modules: HashSet::new(),
+        };
+
+        if loader.vm_config().paranoid_type_checks {
```
Comment (georgemitenkov) [inline third_party/move/move-vm/runtime/src/interpreter.rs:149]: This is beautiful! One thing to note: we also have some paranoid checks inside execution loop (e.g., for functions). These can also be moved to this generic, but maybe you have a separate PR for that?

File: third_party/move/move-vm/runtime/src/interpreter.rs [hunk 12]
```diff
@@ -57,9 +57,21 @@ macro_rules! set_err_info {
 ///
 /// An `Interpreter` instance is a stand alone execution context for a function.
 /// It mimics execution on a single thread, with an call stack and an operand stack.
-pub(crate) struct Interpreter {
+pub(crate) struct Interpreter;
```
Comment (georgemitenkov) [inline third_party/move/move-vm/runtime/src/interpreter.rs:60]: Where is this used? I see only Impl

File: third_party/move/move-vm/runtime/src/debug.rs [hunk 13]
```diff
@@ -102,7 +102,7 @@ impl DebugContext {
         pc: u16,
         instr: &Bytecode,
         resolver: &Resolver,
-        interp: &Interpreter,
+        interp: &dyn InterpreterDebugInterface,
```
[the remark this answers is not in the document]
Comment (ziaptos) [reply third_party/move/move-vm/runtime/src/debug.rs:105]: yes, it was copy-paste, but will update.

File: third_party/move/move-vm/runtime/src/frame_type_cache.rs [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+            std::collections::btree_map::Entry::Occupied(entry) => Ok(entry.into_mut()),
+            std::collections::btree_map::Entry::Vacant(entry) => {
+                let v = ty_func(idx)?;
+                Ok(entry.insert(v))
+            },
+        }
+    }
+
+    #[inline(always)]
+    pub(crate) fn get_field_type_and_struct_type(
```
[the remark this answers is not in the document]
Comment (ziaptos) [reply third_party/move/move-vm/runtime/src/frame_type_cache.rs:55]: mostly

File: third_party/move/move-vm/runtime/src/frame_type_cache.rs [hunk 15]
```diff
@@ -0,0 +1,194 @@
...
+                let struct_type = resolver.field_instantiation_to_struct(idx, ty_args)?;
+                let struct_ty_count = NumTypeNodes::new(struct_type.num_nodes() as u64);
+                let field_ty = resolver.get_generic_field_ty(idx, ty_args)?;
+                let field_ty_count = NumTypeNodes::new(field_ty.num_nodes() as u64);
+                Ok(((field_ty, field_ty_count), (struct_type, struct_ty_count)))
+            })?;
+        Ok(((field_ty, *field_ty_count), (struct_ty, *struct_ty_count)))
+    }
+
+    pub(crate) fn get_variant_field_type_and_struct_type(
```
[the remark this answers is not in the document]
Comment (ziaptos) [reply third_party/move/move-vm/runtime/src/frame_type_cache.rs:72]: We shouldn't be adding that for `consistency`, We should do that to short functions that we actually want inlined. I'll re-evaluate all of them and re-assign the inline flags. For now, I'll keep the originals.

File: third_party/move/move-vm/runtime/src/interpreter.rs (hunk 12 above)
Comment (ziaptos) [reply third_party/move/move-vm/runtime/src/interpreter.rs:60]: That's the public API, I didn't want to change the API, the InterpreterImpl will eventually be generic, while Interpreter can stay as is and do the type-erasure.

File: third_party/move/move-vm/runtime/src/interpreter.rs (hunk 11 above)
Comment (ziaptos) [reply third_party/move/move-vm/runtime/src/interpreter.rs:149]: Exactly - following PRs

File: third_party/move/move-vm/runtime/src/interpreter.rs (hunk 10 above)
Comment (ziaptos) [reply third_party/move/move-vm/runtime/src/interpreter.rs:1157]: Inherited so far, will put on my todo to update

File: third_party/move/move-vm/runtime/src/tracing.rs [hunk 16]
```diff
@@ -6,7 +6,7 @@
 use crate::debug::DebugContext;
 #[cfg(any(debug_assertions, feature = "debugging"))]
 use crate::{
-    interpreter::Interpreter,
+    interpreter::InterpreterImpl,
```
[the remark this answers is not in the document]
Comment (ziaptos) [reply third_party/move/move-vm/runtime/src/tracing.rs]: Yes - good catch, I just missed it. Also seems this is not used as much.

File: third_party/move/move-vm/runtime/src/tracing.rs [hunk 17]
```diff
@@ -72,7 +72,7 @@ pub(crate) fn trace(
     pc: u16,
     instr: &Bytecode,
     resolver: &Resolver,
-    interp: &Interpreter,
+    interp: &InterpreterImpl,
```
[the remark this answers is not in the document]
Comment (ziaptos) [reply third_party/move/move-vm/runtime/src/tracing.rs]: absolutely

File: third_party/move/move-vm/runtime/src/frame_type_cache.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (georgemitenkov) [reply third_party/move/move-vm/runtime/src/frame_type_cache.rs:73]: Fairs. My point was that if we have a reason to force inlining for structs, I do not see how enums should be different. Functions are used in dispatch loop, so inlining should help with perf (I guess that was the original reason). Now that things are in a separate module, inlining might be more important (for privae functions Rust compiler is smart enough). That's why "consistency"

File: third_party/move/move-vm/runtime/src/interpreter.rs (hunk 8 above)
Comment (georgemitenkov) [reply third_party/move/move-vm/runtime/src/interpreter.rs:182]: I meant this is better than `C: ...`

Comment (ziaptos) [reply third_party/move/move-vm/runtime/src/interpreter.rs:182]: Lol, don't know how a "confused" emoji ended up there - wanted to click on a smiley

File: third_party/move/move-vm/runtime/src/interpreter.rs (hunk 12 above)
Comment (georgemitenkov) [reply third_party/move/move-vm/runtime/src/interpreter.rs:60]: Ah, I see, `impl Interpreter` creates `InterpreterImpl`, makes sense!

**Review by georgemitenkov — APPROVED** [review]
LGTM, thanks! Some minor comments from previous review are still pending, but those we can address as a follow-up if you prefer.

File: third_party/move/move-vm/runtime/src/runtime_type_checks.rs (hunk 2 above)
Comment (ziaptos) [reply third_party/move/move-vm/runtime/src/runtime_type_checks.rs:70]: Disagree with this - prefer to have the two choices next to each other (before the implementation). As for documentation - I think the names are self documenting :)

Comment (georgemitenkov) [reply third_party/move/move-vm/runtime/src/runtime_type_checks.rs:70]: Don't have strong opinion on this, so fine with me

**Review by vgao1996 — APPROVED** [review]
LGTM, a step toward the right direction.

I'm aware that the PR mostly moves things around, but given the amount of critical code you've touched, I'd recommend us being cautious and run replay before landing the PR.

**Pull request merged**